gateway/src/main/java/org/georchestra/gateway/security/ResolveGeorchestraUserGlobalFilter.java

```diff
@@ -72,14 +87,28 @@ public class ResolveGeorchestraUserGlobalFilter implements GlobalFilter, Ordered
      * chain.
      */
     public @Override Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
-
-        return exchange.getPrincipal()//
+        Mono<Void> res = exchange.getPrincipal()//
                 .doOnNext(p -> log.debug("resolving user from {}", p.getClass().getName()))//
                 .filter(Authentication.class::isInstance)//
                 .map(Authentication.class::cast)//
-                .map(resolver::resolve)//
+                .map(auth -> {
+                    try {
+                        return Pair.of(resolver.resolve(auth), "");
+                    } catch (DuplicatedEmailFoundException exp) {
+                        Optional<GeorchestraUser> op = Optional.empty();
+                        return Pair.of(op, "/login?error=" + DUPLICATE_ACCOUNT);
+                    }
+                })//
                 .map(user -> {
-                    GeorchestraUser usr = user.orElse(null);
+                    if (!user.getRight().isEmpty()) {
+                        SecurityContextHolder.getContext();
+                        ServerHttpResponse response = exchange.getResponse();
+                        response.setStatusCode(HttpStatus.FOUND);
+                        response.getHeaders().setLocation(URI.create(user.getRight()));
+                        return exchange;
+                    }
+
+                    GeorchestraUser usr = user.getLeft().orElse(null);
```

Comment:
I've tried to better understand Mono and ended up with something like (not fully working) :
```
.map(resolver::resolve)//
                .map(user -> {
                    GeorchestraUser usr = user.orElse(null);
                    GeorchestraUsers.store(exchange, usr);
                    if (usr != null && usr instanceof ExtendedGeorchestraUser) {
                        ExtendedGeorchestraUser eu = (ExtendedGeorchestraUser) usr;
                        Organization org = eu.getOrg();
                        GeorchestraOrganizations.store(exchange, org);
                    }
                    return exchange;
                })//
                .doOnError(DuplicatedEmailFoundException.class, exp -> {
                    ServerHttpResponse response = exchange.getResponse();
                    response.setStatusCode(HttpStatus.FOUND);
                    response.getHeaders().setLocation(URI.create("/login?error=" + DUPLICATE_ACCOUNT));
                    exchange.getSession().flatMap(WebSession::invalidate);
                })
```
I think we can do something way better. I couldn't succeed to handle status code in error.

Reply:
Yes I think we can do something this way, but still have to find the correct syntax. Your proposal run code if an exception is thrown but doesn't cancel the exception which is thrown anyway at the end of the Mono chain. In fact it's like a try-finally but we need instead something like a try-catch.